---
title: "Simulation Report"
author:
    - "Nicholas Knoblauch"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
   rmd: "gen_report.Rmd"
output:
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

## Simulation subset

```{r,echo=F,message=F}

setwd("~/Dropbox/PolygenicRSS/code/snakemake_files/ldsc_pipeline")
library(SeqArray)
library(SeqSupport)
library(tidyverse)
haplo_gds <- seqOpen(snakemake@input[["haplo_gdsf"]])
geno_gds <- seqOpen(snakemake@input[["geno_gdsf"]])

snpinfo <- readRDS(snakemake@input[["snpif"]])
traitn <- snakemake@params[["traitn"]]

p <- calc_p(haplo_gds)
N <- calc_N(geno_gds)

```

The name for this trait is `r traitn`

## Number and distribution of SNPs


For `r traitn`, there are `r p` SNPs. 
SNPs are 

```{r}
group_by(snpinfo,chr) %>% summarise(n_snps=n())  %>% ggplot(aes(x=chr,y=n_snps))+scale_x_discrete(limits=as.character(1:22))+geom_point()
```


## Allele Frequency Spectrum
```{r}

library(ggjoy)
snpinfo <- mutate(
  snpinfo,
  AF=seqAlleleFreq(geno_gds,parallel = 4),
  obs_var= seqBlockApply(geno_gds,"$dosage",partial(apply,MARGIN=2,FUN=var),margin="by.variant",as.is="unlist"),
  exp_var=2*AF*(1-AF)
  )




                  
ggplot(snpinfo,aes(x=AF))+geom_histogram(bins=500)
```

```{r}
ggplot(snpinfo,aes(x=obs_var))+geom_histogram(bins=500)
```
```{r}
ggplot(snpinfo,aes(x=exp_var,y=obs_var))+geom_hex()+geom_abline(intercept = 0,slope = 1)+ggtitle("Observed vs expected variance")
```



```{r}
ggplot(snpinfo,aes(x=exp_var,y=obs_var))+geom_hex()+geom_abline(intercept = 0,slope = 1)+ggtitle("Observed vs expected variance")+facet_wrap(~chr)
```


```{r}
snpinfo  %>% ggplot(aes(x=AF,y=chr))+geom_joy()+scale_y_discrete(limits=as.character(1:22))+scale_x_log10()+ggtitle("Allele Frequency","log10 scale")
```




The size distribution of LD regions (in # of SNPs) looks like this:

```{r}
snpi_ct <- group_by(snpinfo,region_id) %>% summarise(n_snps=n()) %>% arrange(n_snps)
snpi_ct  %>% ggplot(aes(x=n_snps))+geom_histogram(bins=as.integer(min(c(length(unique(snpinfo$region_id))/10,100))))
```

By chromosome it looks like this:

```{r}
library(ggjoy)
snpi_ct <- group_by(snpinfo,chr,region_id) %>% summarise(n_snps=n()) %>% arrange(n_snps)
snpi_ct  %>% ggplot(aes(x=n_snps,y=chr))+geom_joy()+scale_y_discrete(limits=as.character(sort(as.numeric(unique(snpi_ct$chr)))))
```





<!-- ## Source -->
<!-- <a download="report.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">R Markdown source file (to produce this document)</a> -->
