gardner_dir="../../../../../../../../group/xhe-lab/1KG/"
if os.path.exists(gardner_dir):
    configpath="../workflow_params_gardner.json"
else:
    configpath="../workflow_params.json"

configfile:
    configpath

KG_dir="../"+config["KG_DIRECTORY"]

#Directory that holds "raw" 1000 genomes vcf files (I think this already exists in the `compbio` folder in the `RSS` folder)
VCF_DIR="../"+config["VCF_DIR"]

#Directory that holds the haplotype and genotype `.gds` which are population specific (i.e EUR), and have had an allele frequency cutoff applied (i.e 0.05)
GDS_DIR="../"+config["GDS_DIR"]
MGDS_DIR=GDS_DIR+"../all_gds/"

#Directory that holds the 1000 genomes genetic map files (you can get these off github, google 1000 genomes genetic maps)
KG_MAPDIR="../"+config["KG_MAPDIR"]
WTCC_H5_DIR="../"+config["WTCC_H5"]
HDF5_DIR="../"+config["HDF5_DIR"]
# File that has the LDetect LD breakpoints
#URL for that file is here #LD_break_file_url="https://bitbucket.org/nygcresearch/ldetect-data/raw/ac125e47bf7ff3e90be31f278a7b6a61daaba0dc/EUR/fourier_ls-all.bed"
LD_break_file="../"+config["LDETECT_BREAK_FILE"]
OUTPUT_DIR= "../"+config["OUTPUT_DIR"]
INPUT_DIR= "../"+config["INPUT_DIR"]
VCF_DIR="../"+config["VCF_DIR"]
GWAS_NAMES = ["bd","cad","cd","ht","ra","t1d","t2d"]


rule all:
    input:
        expand(HDF5_DIR+"{gn}_seq_hapmap_geno.h5",gn=GWAS_NAMES)



rule subset_wtcc:
    ''' This rule creates the hdf5 file that will be used by most of the subsequent rules, It should be somewhere on RCC or gardner, and shouldn't have to be re-run'''
    input:
        input_gds=MGDS_DIR+"ALL_seq_hapmap_geno.gds",
        input_h5=WTCC_H5_DIR+"combined.h5"
    params:
        gn="{gn}"
    output:
        output_h5=HDF5_DIR+"{gn}_seq_hapmap_geno.h5"
    script:
        "../../scripts/subset_wtcc.R"
