<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Nicholas Knoblauch" />

<meta name="date" content="2017-08-07" />

<title>Comparing RSS-p to ldsc</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">PolygenicRSS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/workflowr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Comparing RSS-p to ldsc</h1>
<h4 class="author"><em>Nicholas Knoblauch</em></h4>
<h4 class="date"><em>2017-08-07</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<p><strong>Last updated:</strong> 2017-09-18</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> 80862d6</p>
<!-- Add your analysis here -->
<div id="setup" class="section level1">
<h1>Setup:</h1>
<p>As a larger test of RSSp, I’ve simulated GWAS data using 1000 genomes EUR individuals (<span class="math inline">\(N=503\)</span>). I’m using the same individuals for the LD panel, which is used by both <code>RSSp</code> and <code>LDSC</code>. As RSSp is still a little slow, I’m only using SNPs from chromosome 2 in this simulation. That works out to 499373 SNPs in total. I am also using the Wen and Stephens estimator (as implemented in <code>LDshrink</code>) to calculate LD (and LD scores). I simulated traits with levels of <span class="math inline">\(PVE\)</span> ranging from 0.01 to 0.9, and confounding (as a proportion of <span class="math inline">\(PVE\)</span>) ranging from 0to 0.09. With 4 replicates of each scenario, there were 108 traits in total.</p>
<div id="estimating-varu" class="section level2">
<h2>Estimating <span class="math inline">\(Var(U)\)</span></h2>
<p><code>ldsc</code>’s version of <span class="math inline">\(Var(U)\)</span> (<span class="math inline">\(\sigma^2_u\)</span>), is <span class="math inline">\(h^2/p\)</span>. <code>RSSp</code> reports <span class="math inline">\(sigma^2_u\)</span> directly</p>
<p><img src="figure/chr_2.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><code>ldsc's</code> estimates are relatively well correlated with <span class="math inline">\(\sigma_u\)</span>, but are all dramatic underestiamtes. Also note that at low <span class="math inline">\(\sigma_u\)</span>, some of <code>ldsc</code>’s estimates are negative.</p>
<p><img src="figure/chr_2.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><code>RSSp</code> seems to slightly overestimate <span class="math inline">\(\sigma^2_u\)</span>, but does a much better job overall.</p>
</div>
<div id="estimating-confounding-interceptinflation" class="section level2">
<h2>Estimating confounding (intercept/inflation)</h2>
<p>Both methods are able to estimate confounding/inflation. <code>ldsc</code> estimates an intercept, which can be positive or negative. <code>RSSp</code> (as currently implemented) estimates a confounding factor <code>c</code> which is (arbitrarily) bounded between <code>0</code> and <code>1</code></p>
<p><img src="figure/chr_2.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><code>ldsc</code> has a hard time estimating confounding when <span class="math inline">\(c\)</span> is low, but does ok as the level of confounding increases.</p>
<p><img src="figure/chr_2.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><code>RSSp</code> consistently understimates the level of confounding, but is much better on average then <code>ldsc</code></p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<!-- Insert the session information into the document -->
<pre><code>R version 3.4.0 (2017-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.3 LTS

Matrix products: default
BLAS/LAPACK: /opt/intel/compilers_and_libraries_2018.0.061/linux/mkl/lib/intel64_lin/libmkl_intel_lp64.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] SeqArray_1.17.4   gdsfmt_1.12.0     reshape2_1.4.2   
 [4] ggjoy_0.3.0       bindrcpp_0.2      LDshrink_1.0     
 [7] dplyr_0.7.2       purrr_0.2.3       readr_1.1.1      
[10] tidyr_0.7.0       tibble_1.3.4      ggplot2_2.2.1    
[13] tidyverse_1.1.1   RSSp_1.0          RcppEigenH5_0.1.0
[16] future_1.6.0     

loaded via a namespace (and not attached):
 [1] httr_1.3.0              jsonlite_1.5           
 [3] modelr_0.1.1            RcppParallel_4.3.20    
 [5] assertthat_0.2.0        stats4_3.4.0           
 [7] GenomeInfoDbData_0.99.0 cellranger_1.1.0       
 [9] yaml_2.1.14             globals_0.10.2         
[11] backports_1.1.0         lattice_0.20-35        
[13] glue_1.1.1              RcppEigen_0.3.3.3.0    
[15] digest_0.6.12           GenomicRanges_1.28.5   
[17] XVector_0.16.0          rvest_0.3.2            
[19] colorspace_1.3-2        htmltools_0.3.6        
[21] Matrix_1.2-11           plyr_1.8.4             
[23] psych_1.7.5             pkgconfig_2.0.1        
[25] broom_0.4.2             listenv_0.6.0          
[27] haven_1.1.0             zlibbioc_1.22.0        
[29] scales_0.5.0            git2r_0.19.0           
[31] IRanges_2.10.3          BiocGenerics_0.22.0    
[33] lazyeval_0.2.0          mnormt_1.5-5           
[35] magrittr_1.5            readxl_1.0.0           
[37] evaluate_0.10.1         nlme_3.1-131           
[39] forcats_0.2.0           xml2_1.1.1             
[41] foreign_0.8-69          tools_3.4.0            
[43] hms_0.3                 stringr_1.2.0          
[45] S4Vectors_0.14.4        munsell_0.4.3          
[47] Biostrings_2.44.2       compiler_3.4.0         
[49] GenomeInfoDb_1.12.2     rlang_0.1.2            
[51] grid_3.4.0              RCurl_1.95-4.8         
[53] bitops_1.0-6            labeling_0.3           
[55] rmarkdown_1.6           gtable_0.2.0           
[57] codetools_0.2-15        R6_2.2.2               
[59] lubridate_1.6.0         knitr_1.17             
[61] bindr_0.1               workflowr_0.7.0        
[63] rprojroot_1.2           stringi_1.1.5          
[65] parallel_3.4.0          Rcpp_0.12.12           
[67] tidyselect_0.1.1       </code></pre>
</div>
</div>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
