<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Nicholas Knoblauch" />

<meta name="date" content="2020-04-15" />

<title>UK Biobank in-sample simulations</title>

<script src="site_libs/header-attrs-2.1/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">PolygenicRSS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/workflowr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<div class="fluid-row" id="header">



<h1 class="title toc-ignore">UK Biobank in-sample simulations</h1>
<h4 class="author">Nicholas Knoblauch</h4>
<h4 class="date">2020-04-15</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-04-15
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>PolygenicRSS/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.4.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed12345code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(12345)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed12345code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(12345)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it difficult for you and others to run your code on a different machine. Change the absolute path(s) below to the suggested relative path(s) to make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
/home/nwknoblauch/Dropbox/PolygenicRSS/workflow/results/sim_ukb/
</td>
<td style="text-align:left;">
../workflow/results/sim_ukb
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomCreRecombinasePolygenicRSStreef1b73655dfcf7c5c4c48f81a1b009d04cdf14dddtargetblankf1b7365a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/CreRecombinase/PolygenicRSS/tree/f1b73655dfcf7c5c4c48f81a1b009d04cdf14ddd" target="_blank">f1b7365</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomCreRecombinasePolygenicRSStreef1b73655dfcf7c5c4c48f81a1b009d04cdf14dddtargetblankf1b7365a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .RData
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .httr-oauth
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/PolygenicEstimation_cache/
    Ignored:    code/.Rhistory
    Ignored:    code/scripts/.Rhistory
    Ignored:    code/snakemake_files/.RData
    Ignored:    code/snakemake_files/.Rhistory
    Ignored:    code/snakemake_files/.snakemake/
    Ignored:    code/snakemake_files/eqtl_check.RDS
    Ignored:    code/snakemake_files/ldsc_est_gwas_genome/
    Ignored:    code/snakemake_files/parameter_generation/
    Ignored:    code/snakemake_files/subset_data_pipeline/
    Ignored:    code/snakemake_files/temp_map_19.RDS
    Ignored:    code/snakemake_files/temp_target_19.RDS
    Ignored:    code/workflow_params.json~
    Ignored:    data/rep_polygenic/

Untracked files:
    Untracked:  _workflowr.yml
    Untracked:  analysis/AFvsp.Rmd
    Untracked:  analysis/ChunkingEffect.Rmd
    Untracked:  analysis/ChunkingLD.Rmd
    Untracked:  analysis/ConfoundingSimulations.Rmd
    Untracked:  analysis/DownSamplingResults.Rmd
    Untracked:  analysis/EVD_SVD.Rmd
    Untracked:  analysis/GREML_RSSp_UKB.Rmd
    Untracked:  analysis/GRM_estimation.Rmd
    Untracked:  analysis/GenomeWideSimulation.Rmd
    Untracked:  analysis/Heritability.Rmd
    Untracked:  analysis/LDshrink_effect.Rmd
    Untracked:  analysis/MeetingSummary1.Rmd
    Untracked:  analysis/Normalization.Rmd
    Untracked:  analysis/Optimiation.org
    Untracked:  analysis/Optimization.Rmd
    Untracked:  analysis/RSSR_PolygenicData.Rmd
    Untracked:  analysis/RSSp.stan
    Untracked:  analysis/RSSp2.stan
    Untracked:  analysis/RSSp_LDshrink.Rmd
    Untracked:  analysis/RSSp_Large_WTCC_noConfounding.Rmd
    Untracked:  analysis/RSSp_Largest_Simulation.Rmd
    Untracked:  analysis/RSSp_WTCC_noConfounding.Rmd
    Untracked:  analysis/RSSp_confounding.Rmd
    Untracked:  analysis/RSSp_ldsc_highPVE_chr1.Rmd
    Untracked:  analysis/RSSp_ldsc_highPVE_hr.Rmd
    Untracked:  analysis/RSSp_ldsc_noConfounding.Rmd
    Untracked:  analysis/Running_ldetect.Rmd
    Untracked:  analysis/SPVE.Rmd
    Untracked:  analysis/SharedSNPAnalysis (dell-desktop's conflicted copy 2018-09-04).Rmd
    Untracked:  analysis/SharedSNPAnalysis.Rmd
    Untracked:  analysis/Summary.Rmd
    Untracked:  analysis/Summary.org
    Untracked:  analysis/Vanilla_LDSC.Rmd
    Untracked:  analysis/chr_16.Rmd
    Untracked:  analysis/derby.log
    Untracked:  analysis/direct_easy.Rmd
    Untracked:  analysis/eQTL_Pipelines.Rmd
    Untracked:  analysis/fquh_mod.RDS
    Untracked:  analysis/importFramExp.Rmd
    Untracked:  analysis/ldmap.org
    Untracked:  analysis/ldshrink_test.Rmd
    Untracked:  analysis/ltximg/
    Untracked:  analysis/nquh_mod.RDS
    Untracked:  analysis/out_res.RDS
    Untracked:  analysis/pvv_opt.Rmd
    Untracked:  analysis/quh_mod.RDS
    Untracked:  analysis/scz_data.org
    Untracked:  analysis/shrinkage_effect.Rmd
    Untracked:  analysis/ukb_dl.Rmd
    Untracked:  analysis/ukb_pve_confounding.Rmd
    Untracked:  code/.directory
    Untracked:  code/.ipynb_checkpoints/
    Untracked:  code/RSSp_abstract.org
    Untracked:  code/RSSp_abstract_notes.org
    Untracked:  code/RSSp_talk_notes.org
    Untracked:  code/RSSp_talk_notes.pdf
    Untracked:  code/RSSp_talk_notes.tex
    Untracked:  code/Untitled.ipynb
    Untracked:  code/Untitled1.ipynb
    Untracked:  code/cpu\scores
    Untracked:  code/dl_1kg.org
    Untracked:  code/ideas.org
    Untracked:  code/ltximg/
    Untracked:  code/params/
    Untracked:  code/quhdypi/
    Untracked:  code/rssp.yml
    Untracked:  code/scikit_allel.org
    Untracked:  code/scripts/#break_df.R#
    Untracked:  code/scripts/#evd_1kg_h5.R#
    Untracked:  code/scripts/#map_uh_covar.R#
    Untracked:  code/scripts/RSSp_ldsc_results.Rmd
    Untracked:  code/scripts/bin/
    Untracked:  code/scripts/ldsc_rssp.Rmd
    Untracked:  code/scripts/matlab2h5.py
    Untracked:  code/scripts/vcf-liftover.sh
    Untracked:  code/scripts/vcf2allel.py
    Untracked:  code/skl_vf.ipynb
    Untracked:  code/snakemake_files/#trait_snakefile#
    Untracked:  code/snakemake_files/EUR.samples
    Untracked:  code/snakemake_files/UKBB_mf.csv
    Untracked:  code/snakemake_files/config.yaml
    Untracked:  code/snakemake_files/environment_rssp.yml
    Untracked:  code/snakemake_files/envs/ldsc.yml
    Untracked:  code/snakemake_files/ldi.RData
    Untracked:  code/snakemake_files/ldsc/
    Untracked:  code/snakemake_files/rssp_ldsc_report/
    Untracked:  code/snakemake_files/sim.RData
    Untracked:  code/snakemake_files/sum.RData
    Untracked:  code/snakemake_files/vert.txt
    Untracked:  code/t_scripts/
    Untracked:  code/test_py.py
    Untracked:  data/GREML_h2.xlsx
    Untracked:  data/LDSC_UKB_SumStatManifest.tsv
    Untracked:  data/NCBI34_to_GRCh38.chain.gz
    Untracked:  data/SNP_ws.RData
    Untracked:  data/Snakemake_inputs/
    Untracked:  data/all_df.RDS
    Untracked:  data/direct_sim_1kg.RDS
    Untracked:  data/journal.pgen.1006711.s002.xlsx
    Untracked:  data/journal.pgen.1006711.s003.xlsx
    Untracked:  data/ldsc_h2.txt.gz
    Untracked:  data/ntr_results.RData
    Untracked:  data/temp_ws.RData
    Untracked:  data/ukb31063_h2_all.02Oct2019.tsv.gz
    Untracked:  docs/#RSSp_WTCC_noConfounding.Rmd#
    Untracked:  docs/RSSp.stan
    Untracked:  docs/RSSp2.stan
    Untracked:  docs/Summary.org
    Untracked:  docs/derby.log
    Untracked:  docs/figure/LDshrink_effect.Rmd/
    Untracked:  docs/figure/RSSp_Large_WTCC_noConfounding.Rmd/
    Untracked:  docs/figure/RSSp_WTCC_noConfounding.Rmd/
    Untracked:  docs/figure/RSSp_ldsc_highPVE_chr1.Rmd/
    Untracked:  docs/figure/RSSp_ldsc_highPVE_hr.Rmd/
    Untracked:  docs/figure/SharedSNPAnalysis.Rmd/
    Untracked:  docs/fquh_mod.RDS
    Untracked:  docs/log4j.spark.log
    Untracked:  docs/ltximg/
    Untracked:  docs/nquh_mod.RDS
    Untracked:  docs/quh_mod.RDS
    Untracked:  docs/scz_data.org
    Untracked:  dosage.scan.RData
    Untracked:  dosage.snp.RData
    Untracked:  foobar.csv
    Untracked:  org/
    Untracked:  output/FRAM/
    Untracked:  output/RSSp_snakemake/
    Untracked:  output/bd/
    Untracked:  output/pve/
    Untracked:  output/rssp_res/
    Untracked:  workflow/

Unstaged changes:
    Modified:   PolygenicRSS.Rproj
    Modified:   analysis/ConfoundingAdj.Rmd
    Modified:   analysis/DownsamplingResults2.Rmd
    Modified:   analysis/Exp_Heritability.Rmd
    Modified:   analysis/IndependenceCheck.Rmd
    Modified:   analysis/Meeting_Notes.Rmd
    Modified:   analysis/New_WholeGenome.Rmd
    Modified:   analysis/OutOfSampleLD.Rmd
    Modified:   analysis/OutOfSampleLD_smallPVE.Rmd
    Modified:   analysis/PolygenicEstimation.Rmd
    Modified:   analysis/RSSP_standard_error.Rmd
    Modified:   analysis/RSSp_Chunksize_Investigation.Rmd
    Modified:   analysis/RSSp_Large_Simulation.Rmd
    Modified:   analysis/RSSp_Posterior.Rmd
    Modified:   analysis/RSSp_Prediction.Rmd
    Modified:   analysis/RSSp_ldsc.Rmd
    Modified:   analysis/RSSp_ldsc_gwas_direct_noConfounding.Rmd
    Modified:   analysis/RSSp_ldsc_highPVE.Rmd
    Modified:   analysis/SimulationPipeline.Rmd
    Modified:   analysis/_site.yml
    Modified:   analysis/about.Rmd
    Modified:   analysis/chr_2.Rmd
    Deleted:    analysis/chunks.R
    Modified:   analysis/index.Rmd
    Modified:   analysis/license.Rmd
    Modified:   analysis/lnzDist.Rmd
    Modified:   analysis/results.Rmd
    Modified:   analysis/simulatingGWAS.Rmd
    Modified:   analysis/simulation.Rmd
    Modified:   analysis/ukb_pve.Rmd
    Modified:   code/TF_LDshrink.ipynb
    Modified:   code/dask_LDshrink.ipynb
    Modified:   code/scripts/evd_1kg_h5.R
    Modified:   code/scripts/summarise_gcta.R
    Modified:   code/scripts/ukb_dl.R
    Modified:   code/snakemake_files/LD_snakefile
    Modified:   code/snakemake_files/Snakefile
    Modified:   code/snakemake_files/converters_snakefile
    Modified:   code/snakemake_files/grm_snakefile
    Modified:   code/snakemake_files/param_snakefile
    Modified:   code/snakemake_files/ukb_snakefile

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/CreRecombinase/PolygenicRSS/blob/f1b73655dfcf7c5c4c48f81a1b009d04cdf14ddd/analysis/ukb_sim.Rmd" target="_blank">f1b7365</a>
</td>
<td>
CreRecombinase
</td>
<td>
2020-04-15
</td>
<td>
workflowr::wflow_publish(“analysis/ukb_sim.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="simulation" class="section level1">
<h1>Simulation</h1>
<div id="samples" class="section level2">
<h2>Samples</h2>
<p>First, I randomly selected 12,000 samples from the UK biobank. I computed a GRM from these 12000, and created a list of 11309 unrelated individuals (GRM cutoff 0.05). I then sampled without replacement to obtain 10000 individuals for my GWAS simulation.</p>
</div>
<div id="causal-variantssimulating-phenotype" class="section level2">
<h2>Causal Variants/Simulating Phenotype</h2>
<p>I used GCTA to simulate broadly polygenic traits from 1,077,146 variants that roughly corresponds to the Hapmap3 SNP list. It should be noted that I did <em>not</em> check to ensure that these variants varied in my subsample. I assume that GCTA drops causal variants that do not vary in the sample. These variants are the variants that the Neale lab used for LD score regression on UK biobank traits.</p>
<p>GCTA uses the additive genetic model:</p>
<p><span class="math display">\[y_j = \sum_i(w_{i,j}u_i) + e_j\]</span> where <span class="math inline">\(w_{i,j} = (x_{i,j} - 2p_i) / \sqrt{2p_i(1 - p_i)}\)</span>, (which is undefined for <span class="math inline">\(p_i=0\)</span> and <span class="math inline">\(p_i=1\)</span>). <span class="math inline">\(e_j\)</span> is the residual effect generated from a normal distribution with a mean of 0 and variance of <span class="math inline">\(\text{Var}(\sum_i(w_{i,j}u_i))(1 / h^2 - 1)\)</span>.</p>
<p>I simulated traits with heritabilities of 0.2, 0.1 and 0.5. For each heritability value ,GCTA simulates <em>one</em> vector of “true” effects, and then simulates 10 replicates from the “true” effect.</p>
</div>
<div id="gwas" class="section level2">
<h2>GWAS</h2>
<p>I used the <code>plink2</code> glm function to fit the gwas, and used 10 principle components (also obtained from <code>plink2</code>) as covariates, and a 5 percent MAF threshold.</p>
</div>
<div id="gcta" class="section level2">
<h2>GCTA</h2>
<p>I estimated individual-level heritability using GCTA’s GREML, again using 10 PCs as continuous covariates.</p>
</div>
<div id="ldsc" class="section level2">
<h2>LDSC</h2>
<p>I estimated LD scores from the 10000 samples, then estimated heritability (with and without an intercept) using those LD scores and the summary statistics.</p>
</div>
<div id="rssp" class="section level2">
<h2>RSSp</h2>
<p>I ran RSSp on the gwas summary stats using the 10000 samples as the reference LD panel.</p>
<p><img src="figure/ukb_sim.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><img src="figure/ukb_sim.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Manjaro Linux

Matrix products: default
BLAS/LAPACK: /usr/lib/libopenblas_haswellp-r0.3.9.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] EigenH5_2.0.7     ldmap_1.5.2       vroom_1.2.0.9000  magrittr_1.5     
 [5] fs_1.3.2.9000     forcats_0.4.0     stringr_1.4.0     dplyr_0.8.99.9002
 [9] purrr_0.3.3.9000  readr_1.3.1       tidyr_1.0.2.9000  tibble_3.0.0     
[13] tidyverse_1.2.1   plotly_4.9.0.9000 ggplot2_3.3.0    

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.4.6          lubridate_1.7.4       lattice_0.20-38      
 [4] assertthat_0.2.1      rprojroot_1.3-2       digest_0.6.25        
 [7] R6_2.4.1              cellranger_1.1.0      backports_1.1.5      
[10] evaluate_0.14         httr_1.4.1            highr_0.8            
[13] pillar_1.4.3.9001     rlang_0.4.5.9000      lazyeval_0.2.2       
[16] readxl_1.3.1          rstudioapi_0.11       data.table_1.12.8    
[19] whisker_0.4           rmarkdown_2.1         labeling_0.3         
[22] htmlwidgets_1.3       bit_1.1-15.2          munsell_0.5.0        
[25] broom_0.5.2           compiler_3.6.3        modelr_0.1.4         
[28] xfun_0.12             pkgconfig_2.0.3       htmltools_0.4.0      
[31] tidyselect_1.0.0.9000 workflowr_1.4.0       fansi_0.4.1          
[34] viridisLite_0.3.0     crayon_1.3.4          withr_2.1.2          
[37] grid_3.6.3            nlme_3.1-144          jsonlite_1.6.1       
[40] gtable_0.3.0          lifecycle_0.2.0.9000  git2r_0.26.1         
[43] scales_1.1.0          RcppParallel_4.4.4    cli_2.0.2            
[46] stringi_1.4.6         farver_2.0.3          xml2_1.2.2           
[49] ellipsis_0.3.0.9000   generics_0.0.2        vctrs_0.2.99.9010    
[52] tools_3.6.3           bit64_0.9-7           glue_1.3.2           
[55] hms_0.5.3             yaml_2.2.1            colorspace_1.4-1     
[58] rvest_0.3.4           knitr_1.28            haven_2.1.0          </code></pre>
</div>
</div>
</div>

<hr>
<p>
    
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
