---
title: "Out Of Sample LD"
author: "Nicholas Knoblauch"
date: 2018-04-12
output: workflowr::wflow_html
---

# Out of Sample LD

For these simulations I'm using the wtccc `cd` genotypes for GWAS simulation and the `bd` genotypes as a reference panel for the purposes of calculating LD.  

##  Simulation details




```{r,echo=F,message=F,warnings=F}
library(EigenH5)
library(tidyverse)
```



```{r,echo=F,message = F,warnings = F}
res_d_rssp <- dir("~/Dropbox/PolygenicRSS/output/RSSp_snakemake", full.names = T, pattern = "SNP")
res_d_ldsc <- dir("~/Dropbox/PolygenicRSS/output/RSSp_snakemake/ldsc_res/", pattern = "cd", full.names = T)



res_df_rssp <- map_df(res_d_rssp, function(x){
    quietly(read_tsv)(x)$result %>%
                           mutate(chrom = as.character(chrom),
                                  method = "RSSp",
                                  useLDetect = as.character(useLDetect)) %>% as.data.frame() %>% as_tibble()
}) %>%
    filter(!is.na(simulation), !is.na(useLDshrink)) %>%
    filter(scenario == "RepPVE",
           dataset == "cd",
           panel_dataset == "bd",
           chrom == "1-22",
           simulation == "gwas",
           !is.na(chrom),
           !is.na(AF)) %>%
    select(-ends_with("bound"),
           -sigu,
           -lnZ,
           -bias,
           -tbias,
           -optim,
           -useGradient,
           -log_params,
           -message,
           -convergence)

res_df_ldsc <- map_df(res_d_ldsc,
                      function(x) {
                          quietly(read_delim)(x, delim = "\t")$result %>%
                                                                 mutate(Intercept = as.character(Intercept),
                                                                        chrom = as.character(chrom),
                                                                        useLDetect = as.character(useLDetect),
                                                                        method = ifelse(useIntercept, "LDSC_Intercept", "LDSC_noIntercept")) %>%
                                                                 as.data.frame() %>%
                                                                 as_tibble()
                      }) %>%
    filter(scenario == "RepPVE",
           dataset == "cd",
           panel_dataset == "bd",
           chrom == "1-22",
           simulation == "gwas",
           !is.na(chrom),
           !is.na(AF)) %>%
    select(-Lambda_GC,
           -Mean_Chi_2,
           -Total_time_elapsed,
           -Intercept)
res_df <- bind_rows(res_df_rssp, res_df_ldsc)


## ggplot(res_df,
##        aes(x=tsigu,y=abs((pve-tpve)/tpve),group=interaction(tsigu,method),fill=method,col=method)) + geom_boxplot()

sum_df <- res_df %>%
    group_by_at(vars(-pve, -fgeneid)) %>%
    summarise(mean_pve = mean(pve), sd_pve = sd(pve), error_pve = mean((pve - tpve)^2), error_sd = sd((pve - tpve) / tpve), rel_error_pve = mean(((pve - tpve) / tpve)^2)) %>%
    ungroup() %>%
    mutate(error_rmse = sqrt(error_pve))
```


```{r,echo=F,message = F,warnings = F}
mp <- filter(res_df,p==max(p)) %>% filter(n==max(n))
filter(res_df,p==max(p)) %>% filter(n==max(n)) %>% mutate(pve_c=as.factor(tpve)) %>% 
    ggplot(aes(x=tpve,y=pve,col=method,group=interaction(tpve,method)))+geom_boxplot(position=position_dodge(width=0.04))+geom_abline(slope=1)+geom_point(position=position_dodge(width=0.04))+
    ggtitle("Out of sample LD", paste0("p = ", unique(mp$p), ",N = ", unique(mp$n))) +
    xlab("PVE") + 
    ylab(bquote(hat(PVE)))
ggsave("../output/rssp_ldsc_oos.png")    

```
```{r,echo=F,message = F,warnings = F}
filter(res_df,p==max(p)) %>% filter(n==max(n)) %>% filter() %>%
    ggplot(aes(x=tpve,y=pve,col=method))+geom_point()+geom_abline(slope=1)+
    ggtitle("LDSC", paste0("p = ", unique(mp$p), ",N = ", unique(mp$n))) +
    xlab("PVE") + 
    ylab(bquote(hat(PVE)))+geom_smooth(method="lm")+
    facet_grid(method~.)
```


```{r,echo=F,message = F,warnings = F}
filter(res_df,p==max(p)) %>% filter(n==max(n)) %>% filter(method=="RSSp") %>%
    ggplot(aes(x=tpve,y=abs(pve-tpve)/tpve))+
    geom_point()+geom_smooth(method="lm")+
    ggtitle("RSSp", paste0("p = ", unique(mp$p), ",N = ", unique(mp$n))) +
    xlab("PVE") + 
    ylab("RMSE")

```
```{r,echo=F,message = F,warnings = F}
filter(res_df,p==max(p)) %>%
    filter(n==max(n)) %>%
    ggplot(aes(x=tpve,y=abs(pve-tpve)/tpve,col=method))+
    geom_point()+geom_smooth(method="lm")+
    ggtitle("LDSC", paste0("p = ", unique(mp$p), ",N = ", unique(mp$n))) +
    xlab("PVE") + 
    ylab("RMSE")+
    facet_grid(method~.)
```





```{r,echo=F,message = F,warnings = F}
mtpve <- unique(sum_df$tpve)
p <- filter(sum_df,SNPCT==0,
            method=="RSSp",n==2000)  %>% mutate(Method=method,use_ldshrink=useLDshrink) %>% 
  ggplot(aes(x = tpve, y = error_rmse,fill=use_ldshrink,col=use_ldshrink)) + facet_grid(AF~.,labeller = "label_both") +
  geom_col(position=position_dodge())+ geom_errorbar(aes(ymin=error_rmse,ymax=error_rmse+error_sd/2),position=position_dodge())+
  ggtitle("ldshrink improves RSSp PVE estimates","10 simulations per PVE value (N=2000)") +
  xlab("PVE") + scale_x_continuous(breaks=mtpve)+
  ylab("RMSE(PVE)")
p
```


```{r,echo=F,message = F,warnings = F}
p <- filter(sum_df,SNPCT==0,N!=2000,
            method!="LDSC_Intercept",((useLDshrink & method=="RSSp") | (!useLDshrink & method !="RSSp")))  %>% mutate(Method=method) %>% 
  ggplot(aes(x = tpve, y = error_rmse,fill=Method,col=Method)) + facet_grid(AF~n,labeller = "label_both") +
  geom_col(position=position_dodge())+ geom_errorbar(aes(ymin=error_rmse,ymax=error_rmse+error_sd/2),position=position_dodge())+
  ggtitle("Root Mean Squared Error of Out-of-Sample-LD-based PVE Estimates","10 replicates per PVE value") +
  xlab("PVE") + 
  ylab("RMSE(PVE)")
p
```



```{r,echo=F,message = F,warnings = F}
p <- filter(sum_df,SNPCT==0,N==0,AF!=0.2) %>% ggplot(aes(x = tpve, y = mean_pve,col=method)) +
   geom_errorbar(aes(ymin=mean_pve-sd_pve/2,ymax=mean_pve+sd_pve/2))+
  geom_abline(slope = 1, intercept = 0) +geom_point()+
  facet_grid( AF~useLDshrink,labeller = "label_both") +
  ggtitle("RSSp vs LDSc across Allele Frequencies (with and without LDshrink)") +
  xlab("PVE") + 
  ylab("PVE Estimate")
ggplotly(p)
```


```{r,echo=F,message = F,warnings = F}
px <- filter(sum_df,SNPCT==0,N==0,AF!=0.2) %>% ggplot(aes(x = method, y = error_pve,fill=method)) +
  geom_boxplot() + geom_point()+
  facet_grid( AF~useLDshrink,labeller = "label_both",scales="free_y") +
  ggtitle("RSSp performance is vastly improved by LDshrink") +
  xlab("Method") + 
  ylab("PVE squared error")+
  scale_y_sqrt()

ggplotly(px)
```

### How much Better is `RSSp` than `LDSC`?


```{r,echo=F,message = F,warnings = F}

filter(sum_df,N==0,SNPCT==0,AF!=0.2) %>% group_by(method,useLDshrink) %>% summarise(mean_error_pve=mean(rel_error_pve))%>% spread(method,mean_error_pve) %>% mutate(percent_improvement_Intercept=(LDSC_Intercept-RSSp)/LDSC_Intercept,percent_improvement_noIntercept=(LDSC_noIntercept-RSSp)/LDSC_noIntercept)

```

```{r}
filter(sum_df,N==0,SNPCT==0,AF!=0.2,method=="RSSp") %>% group_by(useLDshrink) %>% summarise(mean_error_pve=mean(rel_error_pve))%>% mutate(useLDshrink=ifelse(useLDshrink,"LDshrink","noLDshrink"))%>% spread(useLDshrink,mean_error_pve)%>% mutate(percent_improvement=(noLDshrink-LDshrink)/noLDshrink)
```



## Varying SNP count


```{r,echo=F,message = F,warnings = F}
npx <- filter(sum_df,SNPCT!=0,N==0,AF!=0.2,useLDshrink) %>% ggplot(aes(x = tpve, y = mean_pve,col=method)) +
   geom_errorbar(aes(ymin=mean_pve-sd_pve/2,ymax=mean_pve+sd_pve/2))+
  geom_abline(slope = 1, intercept = 0) +geom_point()+
  facet_grid( AF~p,labeller = "label_both") +
  ggtitle("RSSp is robust across AF and SNP count") +
  xlab("PVE") + 
  ylab("PVE Estimate")
ggplotly(npx)
```


```{r,echo=F,message = F,warnings = F}
fpx <- filter(sum_df,SNPCT!=0,N==0,AF!=0.2,useLDshrink)  %>% ggplot(aes(x = method, y = error_pve,fill=method)) +
  geom_boxplot() + 
  facet_grid( AF~p,labeller = "label_both",scales="free_y") +
  ggtitle("RSSp outperforms LDSC across SNP count and Allele Frequency") +
  xlab("Method") + 
  ylab("PVE squared error")+
  scale_y_sqrt()
ggplotly(fpx)
```


## Varying Sample size

I also varied sample size.  In the `N=2000` scenario, both GWAS and reference panel genotypes were downsampled to `2000` individuals.  In the `N=4686` scenario, `N` reflects the size of the GWAS dataset, the reference panel dataset is `4806`.

```{r,echo=F,message = F,warnings = F}

np <- filter(sum_df,SNPCT==0,AF!=0.2,useLDshrink) %>% ggplot(aes(x = tpve, y = mean_pve,col=method)) +
   geom_errorbar(aes(ymin=mean_pve-sd_pve/2,ymax=mean_pve+sd_pve/2))+
  geom_abline(slope = 1, intercept = 0) +geom_point()+
  facet_grid( AF~n,labeller = "label_both") +
  ggtitle("RSSp is robust to changes in sample size") +
  xlab("PVE") + 
  ylab("PVE Estimate")
ggplotly(np)
```


```{r,echo=F,message = F,warnings = F}
nnp <- filter(sum_df,SNPCT==0,AF!=0.2,useLDshrink)  %>% ggplot(aes(x = method, y = error_pve,fill=method)) +
  geom_boxplot() + 
  facet_grid( AF~n,labeller = "label_both",scales="free_y") +
  ggtitle("RSSp outperforms LDSC across sample size and Allele Frequency") +
  xlab("Method") + 
  ylab("PVE squared error")+scale_y_sqrt()
ggplotly(nnp)
```
<!-- # Estimates when PVE is precisely 0 -->

<!-- These (2) simulations were performed using the `RA` dataset -->

<!-- ```{r} -->
<!-- res_d_rssp <- dir("~/Dropbox/PolygenicRSS/output/RSSp_snakemake",pattern="ra",full.names = T) -->
<!-- # res_d_rssp <- dir("~/Dropbox/PolygenicRSS/output/RSSp_snakemake",pattern="ra",full.names = T) -->


<!-- res_df <- map_df(res_d_rssp,read_delim,delim="\t") %>% filter(scenario=="NoPVE")  -->
<!-- tparam_df <- distinct(res_df,fgeneid,tsigu,tpve,p,n) -->
<!-- ggplot(res_df,aes(x=tpve,y=pve))+geom_point()+geom_smooth(method="lm",formula=y~x+0)+facet_grid(useLDshrink~simulation,labeller = "label_both")+ggtitle("RSSp",paste0("p=",unique(tparam_df$p),",N=",unique(tparam_df$n)))+xlab("PVE")+ylab(bquote(hat(PVE))) -->
<!-- # rss_res <- select(res_df,fgeneid,tpve,pve,simulation,useLDshrink) %>% mutate(RSSp=tpve-pve) %>% select(-pve) -->
<!-- ``` -->




<!-- ```{r} -->
<!-- res_d_rssp <- dir("~/Dropbox/PolygenicRSS/output/RSSp_snakemake/ldsc_res/",pattern="ra",full.names = T) -->
<!-- res_df <- map_df(res_d_rssp,~read_delim(.x,delim="\t") %>% mutate(Intercept=as.character(Intercept))) %>% filter(scenario=="NoPVE")  %>% inner_join(tparam_df) -->
<!-- ggplot(res_df,aes(x=tpve,y=pve))+geom_point()+geom_smooth(method="lm",formula=y~x+0)+facet_grid(useLDshrink~useIntercept,labeller = "label_both")+ggtitle("LDSC",paste0("p=",unique(tparam_df$p),",N=",unique(tparam_df$n)))+xlab("PVE")+ylab(bquote(hat(PVE))) -->

<!-- ``` -->

