---
title: "UK Biobank simulations"
author: "Nicholas Knoblauch"
date: "2020-04-15"
output: workflowr::wflow_html
---


# Simulation 

## Samples
First, I randomly selected 12,000 samples from the UK biobank.  I computed a GRM from these 12000, and created a list of 11309 unrelated individuals (GRM cutoff 0.05).  I then sampled without replacement to obtain 10000 individuals for my GWAS simulation.  I repeated the process to obtain 10000 individuals for a reference LD panel.  


## Causal Variants/Simulating Phenotype 
I used the method `simu` to simulate polygenic traits from the~8 million variants that were above 1 percent allele frequency in both the GWAS and reference LD set.  `RSSp` does not accomodate missing data, so 

`simu` uses the GCTA-style additive genetic model:

$$y_j = \sum_i(w_{i,j}u_i) + e_j$$
where $w_{i,j} = (x_{i,j} - 2p_i) / \sqrt{2p_i(1 - p_i)}$, (which is undefined for $p_i=0$ and $p_i=1$). $e_j$ is the residual effect generated from a normal distribution with a mean of 0 and variance of $\text{Var}(\sum_i(w_{i,j}u_i))(1 / h^2 - 1)$.

I simulated traits with 8 heritabilities of 0.1 to 0.8 in increments of 0.1.  For each heritability value I simulated 10 phenotypes.  For each phenotype, a different vector of true effects was generated 

## GWAS 

I used the `GCTA`'s `fastGWA` to fit the gwas, and used 10 principle components (also obtained from `plink2`) as covariates. 

## LD

For both the GWAS samples and the reference LD samples I used LDshrink to estimate LD on variants within LDetect blocks.  I used the 1kg GBR recombination rate estimated using the recently developed [pyrho](https://github.com/popgenmethods/pyrho) method. I also estimated LD within LDetect blocks without using the shrinkage estimator.

## GCTA

I estimated individual-level heritability using GCTA's single-component GREML, using 10 PCs as continuous covariates.

## LDSC

I estimated LD scores from the 10000 samples using `ldsc.py`.  Note that `ldsc.py` uses a sliding window based on genetic map and not LDshrink or LDetect. I used the `pyrho` recombination rate estimates here as well. I  then estimated heritability (with and without an intercept) using those LD scores and the summary statistics.  




## RSSp

I ran RSSp on the gwas summary stats using either the 10000 GWAS samples (in-sample LD) or used the 10000 held-out,reference panel individuals.


```{r,echo=F,message=F,warning=F}
suppressMessages(library(plotly))
suppressMessages(library(tidyverse))
library(fs)
library(forcats)
library(unglue)
library(magrittr)
library(vroom)
library(ldmap)
library(EigenH5)
parse_ldsc <- function(h2lf){
    h2_dat <- suppressMessages(scan(h2lf,what=character(), sep = "\n",quiet=TRUE))
    h2_rowi <- grep("Total Observed scale", h2_dat)
    h2_row <- h2_dat[h2_rowi]
    h2_data <- h2_dat[h2_rowi:length(h2_dat)]
    h2_data <- h2_data[grep("^[^:]+:[^:]+$", h2_data)]
    h2_datd <- purrr::transpose(strsplit(h2_data, split=":"))
    names(h2_datd) <- c("Variable", "Value")
    h2_datdf <- tidyr::unnest(as_data_frame(h2_datd)) %>%
        dplyr::mutate(Variable=chartr(" ^", "__", Variable), Value = trimws(Value)) %>%
      tidyr::separate(Value, c("Est", "SD"), sep = "[ s\\(\\)]+", remove = T, fill = "right", convert = T)
    return(as.numeric(h2_datdf$Est[1]))
    #%>% replace_na(list(Intercept=1.0,Intercept_sd=0.0))
}
parse_rssp <- function(x){
  readRDS(x) %>% pull(pve)
}
parse_gcta <- function(x){
  tdf <- read_tsv(x)
  select(tdf,-SE) %>% mutate(row=1) %>% tidyr::spread(Source,Variance) %>%
    rename(h2=`V(G)/Vp`) %>% mutate(h2_sd=tdf$SE[4]) %>% pull(h2)
}
parse_file <- function(path,model,...){
  if(str_detect(model,"RSSP")){
    parse_rssp(path)
  }else{
    if(str_detect(model,"GCTA")){
      parse_gcta(path)
    }else{
      parse_ldsc(path)
    }
  }
}
```



```{r,echo=FALSE,message=FALSE}
polym_results <- as.character(fs::dir_ls("/home/nwknoblauch/Dropbox/PolygenicRSS/workflow/results/",recurse = TRUE,glob = "*polym*"))

#rss_df <- map_df(poly_results,readRDS)
```


```{r,echo=FALSE,message=FALSE}
polym_df <- bind_rows(
  unglue(polym_results,"/home/nwknoblauch/Dropbox/PolygenicRSS/workflow/results/sim_ukb_{ind}/polym_{rep}_10_{true_h2}_10000.{ext}")) %>% 
  mutate(path=polym_results) %>% filter(ext!="log") %>% 
  mutate(method=case_when(ext=="int.log" ~ "LDSC_INT",
                         ext=="noint.log"~"LDSC_NOINT",
                         ext=="shrink.RDS"~"RSSP",
                         ext=="noshrink.RDS"~"RSSP_NOSHRINK",
                         ext=="hsq"~"GCTA"),
         ind=if_else(ind=="ind","In-sample LD","Out-of-sample LD"))
results_df <- mutate(polym_df,
         h2=map2_dbl(path,method,~quietly(parse_file)(.x,.y)$result),
         true_h2=as.numeric(true_h2)) %>% 
  as_tibble()
         
```


```{r,echo=FALSE,message=FALSE}
library(firatheme)
rename(results_df,`Heritability Estimation\n Method`=method) %>% ggplot(aes(x=true_h2,y=h2,group=interaction(`Heritability Estimation\n Method`,true_h2),fill=`Heritability Estimation\n Method`))+geom_boxplot()+facet_wrap(~ind)+geom_abline(slope=1)+xlab(bquote(h^2))+ylab(bquote(hat(h^2)))+theme_fira()
```

Overall RSSp outperforms LDSC across a broad range of heritability levels, and performs comparably to in-sample methods for heritability estimation.  LDshrink does not seem to consistently or significantly improve performance.  Most  importantly perhaps, it appears that the issue of inflated PVE estimates as the true PVE increases has gone away.  On the contrary, it appears that LDSC (with or without an intercept) is "deflated": the method seems tounderstimate heritability for higher heritability traits in a predictable fashion. 


```{r}
rename(results_df,`Heritability Estimation\n Method`=method) %>% 
ggplot(aes(x=true_h2,y=h2-true_h2,group=interaction(`Heritability Estimation\n Method`,true_h2),fill=`Heritability Estimation\n Method`))+geom_boxplot()+facet_wrap(~ind)+ggtitle("Error in PVE estimation as a function of true PVE")+geom_hline(yintercept = 0)+xlab(bquote(h^2))+ylab(bquote(hat(h^2)-h^2))
```


```{r}
filter(results_df,str_detect(method,"RSSP")) %>%
  ggplot(aes(x=true_h2,y=(h2-true_h2)^2,group=interaction(method,true_h2),fill=method))+geom_boxplot()+facet_wrap(~ind)+ggtitle("Error in PVE estimation as a function of true PVE")+geom_hline(yintercept = 0)
```


If we just look at the two `RSSp` (with and without `LDshrink`), It looks like `RSSp` without `LDshrink` does a little better, but not much. What's more striking is how close the in-sample and out-of-sample LD results are.


```{r,message=FALSE,warning=FALSE}
filter(results_df,str_detect(method,"RSSP")) %>%
  group_by(method,true_h2,ind) %>% 
  summarise(error=mean((h2-true_h2)^2),error_sd=sd((h2-true_h2^2))) %>% ggplot(aes(x=true_h2,y=error,group=interaction(true_h2,ind),fill=ind))+geom_col(position = "dodge")+ylab("average squared error")+facet_wrap(~method)

```



```{r,message=FALSE,warning=FALSE}
filter(results_df,str_detect(method,"RSSP")) %>%
  group_by(method,ind) %>% 
  summarise(error=mean((h2-true_h2)^2),error_sd=sd((h2-true_h2^2))) 
```



